### ISSUE TITLE
Sample code only sends 1 event... then hangs

### ISSUE THREAD

**Issue by labaxter** [dart-lang/sse#63]
I recently closed my previous issue, as I was able to find some dart client code.

Now, all I want to do is, send notification events from my webservice to clients that subscribe to these events. 
If multiple clients subscribe, I want them all to receive whatever event the server sends.

The first attachment is a simple service based on the example code in this package.  It sends out 100 messages at two second intervals, then quits. (rename this to remove ".txt" (github would not accept a .dart file as an attachement).

[server_based_on_sse_sample_server.dart.txt](https://github.com/dart-lang/sse/files/9334472/server_based_on_sse_sample_server.dart.txt)

There are numerous debug print statements in this server.

The second attachment is a command line client that uses the flutter_sse_client package to subscribe to the service and listen for events:  
[flutter_sse_client.dart.txt](https://github.com/dart-lang/sse/files/9334494/flutter_sse_client.dart.txt)

The third attachment is an alternate client that just does the same thing but does not use the flutter_sse_client package:
[alternate_simple_sse_client.dart.txt](https://github.com/dart-lang/sse/files/9334498/alternate_simple_sse_client.dart.txt)

When the server (first attachment) is run, and then either of the two above clients are run, the client only received one message, and then server hangs on `while ( connections.hasNext )` .

Both of the above clients work, when the server is modified, as in the following attachment:

[modified_server.dart.txt](https://github.com/dart-lang/sse/files/9334506/modified_server.dart.txt)

This version of the service will run as expected, until one attempts to connect to it with second instance of either of the two clients (attachment's 2 or 3) while the first one is still connected and receiving messages.  The second instance of the client connects, but never receives any message.  And the server thinks there is only one connection.  I would expect both instances of the client to receive messages while they are listening.

If one then kills the first client which is receiving messages, the second one exits.  It is like they are both trying to listen on the same stream (but first one pulls of the message before the second one can), and when the service closes the connection for the client that was killed, the stream is closed for the second client as well (so it exits).

NOW, restart the first client, this happens in the server:

```
ERROR - 2022-08-14 21:14:36.125036
Asynchronous error
SocketException: Write failed (OS Error: Broken pipe, errno = 32), address = localhost, port = 9999
dart:convert                                   StringConversionSinkMixin.add
package:sse/src/server/sse_handler.dart 98:15  SseConnection._setUpListener

```

The server then continues.

If one subsequently reruns the first client it connects, fails to get any messages, but the server thinks there is a connected client.

At this point, unless  an author or maintainer can suggest what I'm doing incorrectly, I have to conclude this package really doesn't work despite the fact that (when I cloned it), all the tests run. The code in the package seems too complex for me to figure out (quickly anyway), without some conceptual documentation, and some pointers to where to look.  I've spent a week plus on this, so I've taken and additional hour to write up this issue.  I am starting to think I should roll my own publisher.   Oh... and why is there an assumption the that client is writing back to the service.  I though SSE was about "server sent events" (the standard does not to my knowledge require that clients can write back to the server). 

What is the clientId that I find in sse_client.dart?  Where does this come from?  I seems like the a Map of clients is being maintained, but it is not clear how that id is provided? In the initial request I would assume, what header?

**Comment by grouma** [comment]
I'm sorry you are having so much trouble with this package. Comments inline.

> At this point, unless an author or maintainer can suggest what I'm doing incorrectly, I have to conclude this package really doesn't work despite the fact that (when I cloned it), all the tests run. The code in the package seems too complex for me to figure out (quickly anyway), without some conceptual documentation, and some pointers to where to look.

This package is heavily used, but mostly for Google's internal purposes. It's working as intended as far as I know.  The package documentation is a bit light but the README explains the high level intent: 

```
This package is not intended to be a general purpose SSE package, but instead is a bidirectional protocol for use when Websockets are unavailable. That is, both the client and the server expose a sink and stream on which to send and receive messages respectively.
```

I believe the issues you are running into are due to the fact that you are using a client from `package:flutter_client_sse`. That is an unrelated package. The following part of the README explains the issue:

```
Both the server and client have implicit assumptions on each other and therefore a client from this package must be paired with a server from this package.
```

>  I am starting to think I should roll my own publisher. Oh... and why is there an assumption the that client is writing back to the service. I though SSE was about "server sent events" (the standard does not to my knowledge require that clients can write back to the server).

Correct. See the above comment / README. This package is intended to be used in leu of Websockets, hence the bi-directional protocol.

> I am starting to think I should roll my own publisher.

If all you want is basic SSE support, that may be the suitable approach. This package may serve as a good guide.

> What is the clientId that I find in sse_client.dart? Where does this come from?

The `clientId` is a GUID created by the client so that the server can map POST events to the corresponding client.

> I seems like the a Map of clients is being maintained, but it is not clear how that id is provided? In the initial request I would assume, what header?

The client provides it through a request parameter, see [here](https://github.com/dart-lang/sse/blob/00084c43684ddaf7e09c19c5364c4a27eb04efda/lib/client/sse_client.dart#L49).

**Comment by labaxter** [comment]
Thanks for your commentary, and pointers.  I guess I'll roll my own, as you suggest, perhaps, plucking some useful code from your package, as you suggest.  I'm a little surprised you published this as "SSE" given that it is definitely more than that. Maybe it would better have been called `bidirectional_client_server_events` or `bidirectional_sse` instead` of taking the name space of some thing it clearly isn't.   Clearer documentation would have saved me many hours of hacking at it.  Perhaps, a response to my original issue (complaining about the client not compiling in a native dart or flutter environment) would have been helpful.  

Finally thanks for pointing out the client generated and passed UUID in your client code (which I couldn't run natively), I do think it would be useful for the package to provide a native client. Overnight, I was starting to think that there must be client created identifier that was the `clientId`, and if I rolled my own, that I would probably need something like that to maintain a map of client connections; and that the lack of this was probably causing the problems.

**Comment by grouma** [comment]
> I'm a little surprised you published this as "SSE" given that it is definitely more than that.

Yeah it's a bit of a misnomer but deprecating the package and moving to a more descriptively named package is not worth the return on effort.

> Clearer documentation would have saved me many hours of hacking at it. 

I feel the README does a pretty good job explaining the intent and requirements. Feel free to send a PR to clarify further. I'll happily accept.

> Finally thanks for pointing out the client generated and passed UUID in your client code (which I couldn't run natively),

The client code is meant to be run in a browser as the browser contains the necessary logic to receive server sent events.

> I do think it would be useful for the package to provide a native client.

I'm not sure that's feasible, see above.

**Comment by srix55** [comment]
@labaxter a simple glance at the readme would have saved you a ton. That's the first thing most of us do... 😁

**Comment by labaxter** [comment]
@srix55. Not really. I read the readme.  And a good bit of the code, and examples, and tests. I didn't fully understand how it worked until @grouma responded to me in this issue. The limitations where not exactly clear until I had spent some time with it.  I might well have used the package, if there were a non browser client.  I looked at implementing one.  But I still had my head stuck on the idea that it could actually be used for SSE.   Since the client in the package only worked in the browser, I went down the path of looking at other clients. One thing is sure true: the design is NOT clear from the documentation. It it were, I would have understood it truly was NOT SSE, growled, and moved on. I wanted to believe the package could be used what it was called. 

Fundamentally, this package is miss named. It is NOT SSE. @grouma admits as much.  But he didn't insult me with his comments as you just did.  In fact he helped me understand it really was not what I wanted.

>`Note: This package is misnamed. It does not implement Server Sent Events.  It implements Browser (only) Client/Server Bidirectional messaging, where each message from the server must be responded to by the browser (only) client.  It does not have a generic Dart client that will work outside the browser, and there is no intention by build one. 
`

If the above paragraph were in the README I might have wasted a lot less time.  The package really should be renamed, and user code updated.  It is truly unfortunate that this package is named SSE -- it will be the first hit in any search anyone does for 'SSE'.  This is misdirection, plain and simple.  It does not reflect well on the dart community or google. (but only teckies will know -- this is sad, because Dart is so well thought out and has excellent package support generally.)

After @grouma commented saying the `package is heavily used, but mostly for Google's internal purposes.` I realized this was a dead end; and eventually chose to roll my own using generic web sockets, which the `web_socket_channel` and `shelf_web_socket` packages have excellent support for.   

So I now have a true SSE capability where the server publishes events (any string) to a client that connects using a WebSocket connect request. (ws: or wss:).  The service sends events (messages) to all connected clients, implements a keepalive, and closes client connections when the client goes away (after a short time).  In the end it was not a whole lot of code.  I would consider extracting this and publishing it, but it should be called SSE, and that name has been squatted on -- plus I'm not sure I want to be on the hook for maintaining it.
